File: src/main/java/com/example/spark/SparkJob.java
package com.example.spark;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.UUID;

public class SparkJob {

    public static String getTimePartition(long timestamp) {
        // Create a Date object from the timestamp
        Date date = new Date(timestamp);

        // Define the desired format
        SimpleDateFormat sdf = new SimpleDateFormat("'year='yyyy/'month='MM/'day='dd/'hour='HH");

        // Format the Date object into the desired format and return the string
        return sdf.format(date);
    }

    public static void main(String[] args) {
        // Create a Spark configuration
        SparkConf conf = new SparkConf()
                .setAppName("Save UUIDs to HDFS")
                .setMaster("local[*]");
        JavaSparkContext sc = new JavaSparkContext(conf);

        // Generate 10 random UUIDs
        List<String> uuidList = new ArrayList<>();
        for (int i = 0; i < 10; i++) {
            uuidList.add(UUID.randomUUID().toString());
        }

        // Parallelize the UUID list to create an RDD
        JavaRDD<String> uuidRdd = sc.parallelize(uuidList);

        Long executionTime = new Long(System.currentTimeMillis());
        String timePartition = getTimePartition(executionTime);

        uuidRdd.saveAsTextFile("hdfs://hadooop-hadoop-hdfs-nn:9000/spark-job/" + timePartition + "/" + executionTime);

        // Stop the Spark context
        sc.stop();
    }
}

File: pipelines-batch/example/spark-job.py
from pyspark.sql import SparkConf, SparkContext
from datetime import datetime



def get_time_partition(timestamp):
    # Convert the timestamp to a datetime object
    date = datetime.fromtimestamp(timestamp / 1000.0)

    # Format the datetime object into the desired format
    return date.strftime("year=%Y/month=%m/day=%d/hour=%H")


def main():
    # Create a Spark configuration
    conf = SparkConf().setAppName("Save UUIDs to HDFS").setMaster("local[*]")
    sc = SparkContext(conf=conf)

    # Generate 10 random UUIDs
    uuid_list = [str(uuid.uuid4()) for _ in range(10)]

    # Parallelize the UUID list to create an RDD
    uuid_rdd = sc.parallelize(uuid_list)

    # Get the current timestamp
    execution_time = int(datetime.now().timestamp() * 1000)
    time_partition = get_time_partition(execution_time)

    # Save the RDD as a text file to HDFS
    uuid_rdd.saveAsTextFile(f"hdfs://hadooop-hadoop-hdfs-nn:9000/spark-job/{time_partition}/{execution_time}")

    # Stop the Spark context
    sc.stop()


if __name__ == "__main__":
    main()